src/instance_tracker.ts

```diff
@@ -0,0 +1,343 @@
+}
+
+export interface InstanceTrackerOptions {
+    redisClient: Redis.Redis;
+    shutdownManager: ShutdownManager;
+    audit: Audit;
+    idleTTL: number;
+    metricTTL: number;
+    provisioningTTL: number;
+}
+
+export class InstanceTracker {
+    private redisClient: Redis.Redis;
+    private shutdownManager: ShutdownManager;
+    private audit: Audit;
+    private idleTTL: number;
+    private provisioningTTL: number;
+    private metricTTL: number;
+
+    constructor(options: InstanceTrackerOptions) {
+        this.redisClient = options.redisClient;
+        this.shutdownManager = options.shutdownManager;
+        this.audit = options.audit;
+        this.idleTTL = options.idleTTL;
+        this.provisioningTTL = options.provisioningTTL;
+        this.metricTTL = options.metricTTL;
+
+        this.track = this.track.bind(this);
+    }
+
+    instanceKey(details: InstanceDetails, type: string): string {
+        return `instance:${type}:${details.instanceId}`;
+    }
+
+    // @TODO: handle stats for instances
+    async stats(ctx: Context, report: StatsReport): Promise<boolean> {
+        ctx.logger.debug('Received report', { report });
+        const instanceState = <InstanceState>{
+            instanceId: report.instance.instanceId,
+            instanceType: report.instance.instanceType,
+            metadata: <InstanceMetadata>{ ...report.instance },
+            status: {
+                provisioning: false,
+            },
+            timestamp: report.timestamp,
+            shutdownStatus: report.shutdownStatus,
+            shutdownError: report.shutdownError,
+            reconfigureError: report.reconfigureError,
+            statsError: report.statsError,
+        };
+        if (isEmpty(report.stats) || report.statsError) {
+            // empty stats report, so error
+            ctx.logger.error('Empty stats report, not including stats', { report });
+            // TODO: increment stats report error counter
+        } else {
+            let jibriStatusReport: JibriStatusReport;
+            switch (report.instance.instanceType) {
+                case 'jibri':
+                    jibriStatusReport = <JibriStatusReport>report.stats;
+                    instanceState.status.jibriStatus = jibriStatusReport.status;
+                    break;
+                case 'JVB':
+                    instanceState.status.jvbStatus = <JVBStatus>report.stats;
+                    break;
+            }
+        }
+        ctx.logger.debug('Tracking instance state', { instanceState });
+        return await this.track(ctx, instanceState);
+    }
+
+    async track(ctx: Context, state: InstanceState): Promise<boolean> {
+        let group = 'default';
+        // pull the group from metadata if provided
+        if (state.metadata && state.metadata.group) {
+            group = state.metadata.group;
+        }
+
+        // Store latest instance status
+        const key = `instance:cstatus:${group}:${state.instanceId}`;
```

Comment:
I is a typo, or you renamed the `instance:status` prefix into `instance:cstatus`?